pkg/util/util.go

```diff
@@ -785,11 +793,46 @@ func GetGitHubZipURL(repoURL string) (string, error) {
 		repo = strings.TrimSuffix(repo, ".git")
 	}
 
-	// TODO: pass branch or tag from devfile
-	branch := "master"
+	// Extract reference checkout for project
+	references := make(map[string]string)
```

Comment:
Why not a struct instead to store this information? Or just using `project.Source.`?

Reply:
See: https://github.com/openshift/odo/pull/3100#discussion_r420882602 for the struct. We would need to have a few if statements, and probably some boolean to track with permutation is set to then print a nice message for the user. Maybe we can use reflect? That would give us a map of field to value, which is what I created manually. Not sure what is best. By `project.Source.` do you mean creating a var, and setting it depending on the which field is set in `project.Source`?  I did think of something like
```
var ref string
var errMsg
if project.Source.branch != nil {
   ref = project.Source.branch
}

if project.Source.tag != nil {
   if ref != nil {
       ref = project.Source.tag
   } else {
      errMsg += "tag also provided"
      // Issue with error message, and not knowing what ref was set with. Either tag or branch
   }
}
```
Please let me know if I am missing something!

Reply:
I just pushed another commit changing the implementation to booleans, instead of using a map. Please let me know what you think @cdrage  @prietyc123

Reply:
>I just pushed another commit changing the implementation to booleans, instead of using a map. Please let me know what you think @cdrage @prietyc123

I think too many nested if else statement making it very complicated and clumsy. IMO taking `reflect` approach would be a way simpler than this.
You can verify the `Branch` and `tags` also using switch statement.

Reply:
I think the reflect way, would be the same output as the reference map since with reflect we will just be extracting the values of each field and populating a type.